salt/transport/ws.py

```diff
@@ -0,0 +1,605 @@
+        self,
+        port=None,
+        connect_callback=None,
+        disconnect_callback=None,
+        timeout=None,
+    ):
+        if port is not None:
+            self.port = port
+        if connect_callback:
+            self.connect_callback = None
+        if disconnect_callback:
+            self.disconnect_callback = None
+        await self._connect(timeout=timeout)
+
+    async def send(self, msg):
+        await self.message_client.send(msg, reply=False)
+
+    async def recv(self, timeout=None):
+        while self._ws is None:
+            await self.connect()
+            await asyncio.sleep(0.001)
+        if timeout == 0:
+            try:
+                raw_msg = await asyncio.wait_for(self._ws.receive(), 0.0001)
+            except TimeoutError:
+                return
+            if raw_msg.type == aiohttp.WSMsgType.TEXT:
+                if raw_msg.data == "close":
+                    await self._ws.close()
+            if raw_msg.type == aiohttp.WSMsgType.BINARY:
+                return salt.payload.loads(raw_msg.data, raw=True)
+            elif raw_msg.type == aiohttp.WSMsgType.ERROR:
+                log.error(
+                    "ws connection closed with exception %s", self._ws.exception()
+                )
+        elif timeout:
+            return await asyncio.wait_for(self.recv(), timeout=timeout)
+        else:
+            while True:
+                raw_msg = await self._ws.receive()
+                if raw_msg.type == aiohttp.WSMsgType.TEXT:
+                    if raw_msg.data == "close":
+                        await self._ws.close()
+                if raw_msg.type == aiohttp.WSMsgType.BINARY:
+                    return salt.payload.loads(raw_msg.data, raw=True)
+                elif raw_msg.type == aiohttp.WSMsgType.ERROR:
+                    log.error(
+                        "ws connection closed with exception %s",
+                        self._ws.exception(),
+                    )
+
+    async def on_recv_handler(self, callback):
+        while not self._ws:
+            await asyncio.sleep(0.003)
+        while True:
+            msg = await self.recv()
+            await callback(msg)
+
+    def on_recv(self, callback):
+        """
+        Register a callback for received messages (that we didn't initiate)
+        """
+        if self.on_recv_task:
+            # XXX: We are not awaiting this canceled task. This still needs to
+            # be addressed.
+            self.on_recv_task.cancel()
+        if callback is None:
+            self.on_recv_task = None
+        else:
+            self.on_recv_task = asyncio.create_task(self.on_recv_handler(callback))
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+
+class PublishServer(salt.transport.base.DaemonizedPublishServer):
+    """ """
+
+    # TODO: opts!
+    # Based on default used in tornado.netutil.bind_sockets()
+    backlog = 128
```

Comment:
wondering about use of Magic Number 128, what is it, and would it be better defined as some definition name set to 128 ?  Note it's use / setting in unchanged code, but wondering what it represents.

Cannot get the C/C++ and dislike of magic numbers out of me :smile:

Reply:
This represents the number of outstanding connections the socket will listen too and is a hold over from the TCP transport. This value has been in-place for some time. Unless it's causing problems I don't see the need to change it for now.